lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2974,6 +2974,20 @@ ConformanceChecker::inferTypeWitnessesViaValueWitnesses(ValueDecl *req) {
                                          result.second->getCanonicalType()})
                              .second)
                          return true;
+                       
+                       // Filter out circular possibilities, e.g. that
+                       // AssocType == S.AssocType or
+                       // AssocType == Foo<S.AssocType>.
+                       if (result.second.findIf([&](Type t) -> bool {
+                         if (auto dmt = t->getAs<DependentMemberType>()) {
+                           if (dmt->getAssocType()->getName() == result.first->getName()
+                             && dmt->getBase()->isEqual(Conformance->getType()))
```

Comment:
Yeah, it looks like the type we have at this point is the result of trying to substitute `Self == ConformingType`, and then failing to collapse DMTs resulting from substituting `Self.AssocType`. I should test that this works with generic types, though…